### PR TITLE
Rework erc20holder to support all tokens

### PR DESCRIPTION
**Pull request by ThunderDeliverer**
# Description

The ERC-20 token holder has been reworked to support all token types. The name has been changed to RMRKSecureTokenTransferProtocol, which is still a work-in-progress name.

We need to agree on the name before merging this one.

I removed the ability to transfer tokens from one token to another while adding support for ERC-1155 and ERC-721. It seemed that we could achieve similar results by using this to extend ERC-6059 and remove the tokens that we didn't want to transfer with it before the transfer. Once I started addressing the token-to-token transfer of assets, I remembered that we would probably be asked to add multi-token transfer, and at that point, it seemed smarter to transfer the holder token rather than the assets held within it (less internal transactions => less gas consumed => lower transaction cost).

# Checklist

- [x] Verified code additions
- [x] Updated NatSpec comments (if applicable)
- [x] Regenerated docs
- [x] Ran prettier
- [x] Added tests to fully cover the changes

---

### Detailed summary
- Deleted `console.md` file in `docs` directory.
- Deleted `erc20holder.ts` file in `test/extensions` directory.
- Deleted `ERC20Holder.md` and `IERC20Holder.md` files in `docs/RMRK/extension/erc20Holder` directory.
- Deleted `ERC20Holder.sol` and `IERC20Holder.sol` files in `contracts/RMRK/extension/erc20Holder` directory.
- Deleted `RMRKERC20HolderMock.sol` file in `contracts/mocks/extensions/erc20holder` directory.
- Changed the value of `IERC20Holder` constant to `IRMRKTokenHolder`.
- Added `ERC1155Mock.sol` file in `contracts/mocks` directory.
- Added `RMRKTokenHolderMock.sol` file in `contracts/mocks/extensions/tokenHolder` directory.
- Added `IERC1155Receiver.md` file in `docs/elin/contracts/token/ERC1155` directory.
- Added `IRMRKTokenHolder.sol` file in `contracts/RMRK/extension/tokenHolder` directory.
- Modified the `ERC1155.md` file in `docs/elin/contracts/token/ERC1155` directory.

> The following files were skipped due to too many changes: `docs/elin/contracts/token/ERC1155/ERC1155.md`, `docs/mocks/ERC1155Mock.md`, `docs/RMRK/extension/tokenHolder/IRMRKTokenHolder.md`, `docs/elin/contracts/token/ERC1155/IERC1155.md`, `docs/elin/contracts/token/ERC1155/extensions/IERC1155MetadataURI.md`, `docs/RMRK/extension/tokenHolder/RMRKTokenHolder.md`, `contracts/RMRK/extension/tokenHolder/RMRKTokenHolder.sol`, `docs/mocks/extensions/erc20holder/RMRKERC20HolderMock.md`, `test/extensions/tokenHolder.ts`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

**Comment by ThunderDeliverer** [comment]
@steven2308 I'll add & fix the tests in the evening, but I think you can take a look at the smart contracts already 😁

File: contracts/RMRK/extension/secureTokenTransferProtocol/IRMRKSecureTokenTransferProtocol.sol [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+     * @dev If the token type is `ERC-20`, the `heldTokenId` MUST be ignored.
+     * @dev If the token type is `ERC-721`, the `amount` MUST be ignored.
+     * @param tokenContract The address of the token smart contract
+     * @param tokenType The type of the token being transferred
+     * @param tokenId The ID of the token to transfer the tokens to
+     * @param heldTokenId The ID of the held token to transfer
+     * @param amount The number of ERC-20 tokens to transfer
+     * @param data Additional data with no specified format, to allow for custom logic
+     */
+    function transferHeldTokenToToken(
```
Comment (steven2308) [inline contracts/RMRK/extension/secureTokenTransferProtocol/IRMRKSecureTokenTransferProtocol.sol:108]: What a wonderful name. Makes it very evident.

File: contracts/RMRK/extension/secureTokenTransferProtocol/RMRKSecureTokenTransferProtocol.sol [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+    function _transferHeldTokenFromToken(
+        address tokenContract,
+        TokenType tokenType,
+        uint256 tokenId,
+        uint256 heldTokenId,
+        uint256 amount,
+        address to,
+        bytes memory data
+    ) internal {
+        if (amount == 0 && tokenType != TokenType.ERC721) {
```
Comment (steven2308) [inline contracts/RMRK/extension/secureTokenTransferProtocol/RMRKSecureTokenTransferProtocol.sol:60]: Wouldn't it make more sense for amount to be always one for 721s?

Comment (steven2308) [reply contracts/RMRK/extension/secureTokenTransferProtocol/RMRKSecureTokenTransferProtocol.sol:60]: Actually it has to be 1, you are increasing the balance by 1 when receiving but decreasing by zero when sending.

**Review by steven2308 — APPROVED** [review]
This looks really good for a generic token holder, besides the minor (I think) issue.

I would honestly still prefer to have a version which only cares about 20s though. Extensions should be more granular IMO.

File: contracts/RMRK/extension/secureTokenTransferProtocol/IRMRKSecureTokenTransferProtocol.sol (hunk 1 above)
Comment (ThunderDeliverer) [reply contracts/RMRK/extension/secureTokenTransferProtocol/IRMRKSecureTokenTransferProtocol.sol:108]: It feels a bit clunky to me in some places, but I think it's clear about which token is referred to where 😅

File: contracts/RMRK/extension/secureTokenTransferProtocol/RMRKSecureTokenTransferProtocol.sol (hunk 2 above)
Comment (ThunderDeliverer) [reply contracts/RMRK/extension/secureTokenTransferProtocol/RMRKSecureTokenTransferProtocol.sol:60]: Yeah, you are correct. This didn't conform to the parameters that needed to be ignored based on the token type. I fixed it in the most recent commit.

**Review by steven2308 — APPROVED** [review]

**Comment by ThunderDeliverer** [comment]
The latest commit also shaves off some size:

````
 ·-----------------------------------------------|--------------|----------------·
 |  Contract Name                                ·  Size (KiB)  ·  Change (KiB)  │
 ················································|··············|·················
 |  RMRKSecureTokenTransferProtocolMock          ·       7.650  ·        -0.421  │
 ················································|··············|·················
````

**Review by steven2308 — APPROVED** [review]
Great changes.
I would call this simply TokenHolder

**Comment by ThunderDeliverer** [comment]
I was trying to get as close to the EIP name as possible. Maybe having an initial title would help here as well.. 🤔

**Comment by steven2308** [comment]
> I was trying to get as close to the EIP name as possible. Maybe having an initial title would help here as well.. thinking

In that case, I think "Secure" is redundant, anything you publish as an EIP should be Secure, shouldn't it?
Also, "Protocol" seems too much of a word for this case. Finally, "Transfer" does not clearly show the intention of this extension, which is to "Hold" or better yet: "Receive, Hold and Send". Maybe "Hold and Transfer" as Transfer could imply receiving and sending, but I think the Hold part is crucial.

**Comment by ThunderDeliverer** [comment]
> In that case, I think "Secure" is redundant, anything you publish as an EIP should be Secure, shouldn't it?
> Also, "Protocol" seems too much of a word for this case. Finally, "Transfer" does not clearly show the intention of this extension, which is to "Hold" or better yet: "Receive, Hold and Send". Maybe "Hold and Transfer" as Transfer could imply receiving and sending, but I think the Hold part is crucial.

I just wanted to have a proposal which we can abbreviate as `STTP` 😂 

How do you feel about `One token to rule them all`? 💍 

Maybe something along the line of `Universal Token Aggregator and Distributor`?

**Comment by steven2308** [comment]
> Maybe something along the line of `Universal Token Aggregator and Distributor`?

That would work if this was a repo, but it's an extension so I don't like neither Universal nor Aggregator. I really think token holder does it, it clearly expresses the point. Let's get some ideas from team!

**Comment by ThunderDeliverer** [comment]
I think I can rename it to TokenHolder & merge this, and we can rename once more it if needed later 🤔

**Pull request merged**